src/core/geometry/qgsreferencedgeometry.cpp

```diff
@@ -66,3 +68,32 @@ QgsReferencedGeometry QgsReferencedGeometry::fromReferencedRect( const QgsRefere
   return QgsReferencedGeometry( QgsGeometry::fromRect( rectangle ), rectangle.crs() );
 }
 
+QgsReferencedGeometry QgsReferencedGeometry::fromEwkt( const QString &ewkt )
```

Comment:
@nyalldawson FYI, I think that's the thing that's interesting in here

Reply:
Ah yeah my comment was a mistake -- I didn't realise some of the changes were collapsed.

One question I do have with ewkt is if the crs number is a postgres srs id or is always an epsg code. The postgis docs suggest it's a postgis srid, which means the actual crs is server dependent and will vary server by server .. which means that interpretation has to reside in the data provider.

Reply:
Is doing this in the provider not an incompatible change? We suddenly have different data types for columns (geometry or referenced geometry instead of string) and will break project configurations.

Reply:
Right. I guess you'd need to ask the layer's data provider to convert the ewkt to a referenced geometry for you.

Reply:
How would that be requested?

Reply:
Something like `virtual QgsReferencedGeometry QgsVectorDataProvider::fromEWkt( const QString& ewkt )`? Returns null geometry by default, postgres provider can implement with correct crs retrieval.

Reply:
Ah yes, that sounds easier than what I had in mind with an additional feature request flag.

Reply:
Not to forget, we also need a toEwkt method. Or support storing attributes with referenced geometry

Reply:
Actually on second thoughts I'd avoid specifically referencing "ewkt" in the name, and go with "stringToReferencedGeometry" instead. Ewkt is postgis specific, but other providers may have an alternative format for string representation of geom + crs.

Reply:
> Something like `virtual QgsReferencedGeometry QgsVectorDataProvider::fromEWkt( const QString& ewkt )`? Returns null geometry by default, postgres provider can implement with correct crs retrieval.

These seems a cleaner solution to me, I'll go for that. Thank you both for your comments!

Reply:
I was trying to move the functions to the data provider but I can't find a way to access the provider in QgsPalLabeling. Is it not available on purpose? How could i get it?

Reply:
Hm, you're right. The PAL label registration methods can't have any access to the layer, as they'll be used in a background thread. 

It's going to get super complex to handle this. I think the only way would be some small abstract "text to geometry" converter object which can be set in the render context, where you retrieve an instance of that from the layer's data provider and set it in the render context during the map render job preparation stages. (With care taken to make sure it's safe to do the conversion in a different thread from what this object was created in).